### ISSUE TITLE
System.Diagnostics.Tracing.EventParameterInfo.GetMetadataLength() throws NotSupportedException

### ISSUE THREAD

**Issue by mjsabby** [dotnet/coreclr#21743]
After upgrading my application a daily build of .NET Core 3.0 I'm now seeing this exception being thrown quite all over the place.

```+system.net.http!HttpClient.SendAsync
+system.net.http!HttpClient.SendAsync
+system.net.http!HttpMessageInvoker.SendAsync
+system.net.http!HttpClientHandler.SendAsync
+system.net.http!DiagnosticsHandler.SendAsync
+system.private.corelib!System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1[System.__Canon].Start(!!0&)
+system.private.corelib!AsyncMethodBuilderCore.Start
+system.net.http!System.Net.Http.DiagnosticsHandler+<SendAsync>d__2.MoveNext()
+system.diagnostics.diagnosticsource!DiagnosticListener.Write
+system.diagnostics.diagnosticsource!System.Diagnostics.Activity+<get_Baggage>d__24.System.Collections.IEnumerable.GetEnumerator()
+system.diagnostics.diagnosticsource!System.Diagnostics.DiagnosticSourceEventSource+FilterAndTransform+<>c__DisplayClass2_1.<.ctor>b__2(System.Collections.Generic.KeyValuePair`2)
+system.diagnostics.diagnosticsource!DiagnosticSourceEventSource.Activity2Start
+system.private.corelib!EventSource.WriteEventVarargs
+system.private.corelib!EventSource.WriteMultiMergeInner
+system.private.corelib!NameInfo.GetOrCreateEventHandle
+system.private.corelib!EventPipeMetadataGenerator.GenerateMetadata
+system.private.corelib!EventParameterInfo.GetMetadataLength
+coreclr!IL_Throw
+coreclr!RaiseTheExceptionInternalOnly
+kernelbase!RaiseException
+ntdll!RtlRaiseException
+ntdll!RtlDispatchException
+ntdll!RtlpExecuteHandlerForException
+coreclr!ProcessCLRException
+coreclr!ExceptionTracker::ProcessOSExceptionNotification
+coreclr!ExceptionTracker::ProcessManagedCallFrame
+coreclr!??ETW::ExceptionLog::ExceptionThrown
+coreclr!FireEtwExceptionThrown_V1
+coreclr!McTemplateCoU0zzpqhh
+coreclr!EtwCallout
+coreclr!ETW::SamplingLog::SendStackTrace
+EventDataExceptionTypeSystem.NotSupportedException
+EventDataExceptionMessageSpecifiedmethodisnotsupported.
+EventMicrosoft-Windows-DotNETRuntime/Exception/Start

**Comment by mjsabby** [comment]
@brianrob @noahfalk

**Comment by brianrob** [comment]
This is likely caused by an argument type that EventPipe doesn't know how to serialize.  The right behavior here is probably just to handle the failure by not generating metadata for any more arguments for that event.  This would mean that only arguments up to this one would be readable by a generic parser, but you'd not see a failure at runtime.  Note that no data would be lost, so it would still be possible to write a parser that reads the data - it's just that a generic parser that only uses in-trace metadata would not know about the arguments.

**Comment by benaadams** [comment]
@onovotny were you seeing something like this?

**Comment by onovotny** [comment]
I’m seeing it on 2.2 and it’s blocking my updating to it.

**Comment by noahfalk** [comment]
@jorive @tommcdon

**Comment by jorive** [comment]
@mjsabby Do you have a small repro case?

**Comment by onovotny** [comment]
Small, no, but it repro's constantly with my signing service. App Insights seems to call an API that triggers this.

**Comment by sywhang** [comment]
@onovotny Does your project have an ability to take a private runtime that's not published via nuget? I can work on a fix that implements what Brian described above, but without repro I can't verify that it actually works... If not I can try to create a repro this error myself.

**Comment by onovotny** [comment]
@sywhang how would I do that? 

Are you able to repro it by running https://github.com/onovotny/SignService? It repros for me without needing any of the configuration; just set the IIS Express profile as the launch profile and debug.

it breaks here:

```
 	System.Private.CoreLib.dll!System.Diagnostics.Tracing.EventParameterInfo.GenerateMetadataForProperty(System.Diagnostics.Tracing.PropertyAnalysis property, byte* pMetadataBlob, ref uint offset, uint blobSize)	Unknown
 	System.Private.CoreLib.dll!System.Diagnostics.Tracing.EventParameterInfo.GenerateMetadata(byte* pMetadataBlob, ref uint offset, uint blobSize)	Unknown
 	System.Private.CoreLib.dll!System.Diagnostics.Tracing.EventPipeMetadataGenerator.GenerateMetadata(int eventId, string eventName, long keywords, uint level, uint version, System.Diagnostics.Tracing.EventParameterInfo[] parameters)	Unknown
 	System.Private.CoreLib.dll!System.Diagnostics.Tracing.NameInfo.GetOrCreateEventHandle(System.Diagnostics.Tracing.EventProvider provider, System.Collections.Concurrent.ConcurrentDictionary<int, System.IntPtr> eventHandleMap, System.Diagnostics.Tracing.EventDescriptor descriptor, System.Diagnostics.Tracing.TraceLoggingEventTypes eventTypes)	Unknown
 	System.Private.CoreLib.dll!System.Diagnostics.Tracing.EventSource.WriteImpl(string eventName, ref System.Diagnostics.Tracing.EventSourceOptions options, object data, System.Guid* pActivityId, System.Guid* pRelatedActivityId, System.Diagnostics.Tracing.TraceLoggingEventTypes eventTypes)	Unknown
 	System.Private.CoreLib.dll!System.Diagnostics.Tracing.EventSource.Write<<>f__AnonymousType0<string, System.Collections.Generic.IDictionary<string, string>, Microsoft.ApplicationInsights.Extensibility.Implementation.External.EventData, long>>(string eventName, System.Diagnostics.Tracing.EventSourceOptions options, <>f__AnonymousType0<string, System.Collections.Generic.IDictionary<string, string>, Microsoft.ApplicationInsights.Extensibility.Implementation.External.EventData, long> data)	Unknown
 	Microsoft.ApplicationInsights.dll!Microsoft.ApplicationInsights.Extensibility.Implementation.RichPayloadEventSource.WriteEvent<Microsoft.ApplicationInsights.Extensibility.Implementation.External.EventData>(string eventName, string instrumentationKey, System.Collections.Generic.IDictionary<string, string> tags, Microsoft.ApplicationInsights.Extensibility.Implementation.External.EventData data, long flags, System.Diagnostics.Tracing.EventKeywords keywords)	Unknown
 	Microsoft.ApplicationInsights.dll!Microsoft.ApplicationInsights.Extensibility.Implementation.RichPayloadEventSource.Process(Microsoft.ApplicationInsights.Channel.ITelemetry item)	Unknown
 	Microsoft.ApplicationInsights.SnapshotCollector.dll!Microsoft.ApplicationInsights.SnapshotCollector.TelemetryTracker.TrackCustomEvent(string eventName, System.Collections.Generic.IDictionary<string, string> properties, System.Collections.Generic.IDictionary<string, double> metrics)	Unknown
 	Microsoft.ApplicationInsights.SnapshotCollector.dll!Microsoft.ApplicationInsights.SnapshotCollector.SnapshotCollectorTelemetryProcessor.SnapshotCollectorTelemetryProcessor(Microsoft.ApplicationInsights.Extensibility.ITelemetryProcessor nextTelemetryProcessor, Microsoft.ApplicationInsights.SnapshotCollector.IPlatformServices platformServices, Microsoft.ApplicationInsights.SnapshotCollector.SnapshotCollectorConfiguration configuration)	Unknown
>	SignService.dll!SignService.Startup.SnapshotCollectorTelemetryProcessorFactory.Create(Microsoft.ApplicationInsights.Extensibility.ITelemetryProcessor next) Line 207	C#
 	Microsoft.ApplicationInsights.dll!Microsoft.ApplicationInsights.Extensibility.Implementation.TelemetryProcessorChainBuilder.Build()	Unknown
 	Microsoft.ApplicationInsights.AspNetCore.dll!Microsoft.Extensions.DependencyInjection.TelemetryConfigurationOptionsSetup.Configure(Microsoft.ApplicationInsights.Extensibility.TelemetryConfiguration configuration)	Unknown
 	Microsoft.ApplicationInsights.AspNetCore.dll!Microsoft.Extensions.DependencyInjection.TelemetryConfigurationOptions.TelemetryConfigurationOptions(System.Collections.Generic.IEnumerable<Microsoft.Extensions.Options.IConfigureOptions<Microsoft.ApplicationInsights.Extensibility.TelemetryConfiguration>> configureOptions)	Unknown
 	[Native to Managed Transition]	
 	[Managed to Native Transition]	
 	Microsoft.Extensions.DependencyInjection.dll!Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitConstructor(Microsoft.Extensions.DependencyInjection.ServiceLookup.ConstructorCallSite constructorCallSite, Microsoft.Extensions.DependencyInjection.ServiceLookup.ServiceProviderEngineScope scope)	Unknown
 	Microsoft.Extensions.DependencyInjection.dll!Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteVisitor<Microsoft.Extensions.DependencyInjection.ServiceLookup.ServiceProviderEngineScope, object>.VisitCallSite(Microsoft.Extensions.DependencyInjection.ServiceLookup.IServiceCallSite callSite, Microsoft.Extensions.DependencyInjection.ServiceLookup.ServiceProviderEngineScope argument)	Unknown
 	Microsoft.Extensions.DependencyInjection.dll!Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitScoped(Microsoft.Extensions.DependencyInjection.ServiceLookup.ScopedCallSite scopedCallSite, Microsoft.Extensions.DependencyInjection.ServiceLookup.ServiceProviderEngineScope scope)	Unknown
 	Microsoft.Extensions.DependencyInjection.dll!Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitSingleton(Microsoft.Extensions.DependencyInjection.ServiceLookup.SingletonCallSite singletonCallSite, Microsoft.Extensions.DependencyInjection.ServiceLookup.ServiceProviderEngineScope scope)	Unknown
 	Microsoft.Extensions.DependencyInjection.dll!Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteVisitor<Microsoft.Extensions.DependencyInjection.ServiceLookup.ServiceProviderEngineScope, object>.VisitCallSite(Microsoft.Extensions.DependencyInjection.ServiceLookup.IServiceCallSite callSite, Microsoft.Extensions.DependencyInjection.ServiceLookup.ServiceProviderEngineScope argument)	Unknown
 	Microsoft.Extensions.DependencyInjection.dll!Microsoft.Extensions.DependencyInjection.ServiceLookup.DynamicServiceProviderEngine.RealizeService.AnonymousMethod__0(Microsoft.Extensions.DependencyInjection.ServiceLookup.ServiceProviderEngineScope scope)	Unknown
 	Microsoft.Extensions.DependencyInjection.dll!Microsoft.Extensions.DependencyInjection.ServiceLookup.ServiceProviderEngine.GetService(System.Type serviceType, Microsoft.Extensions.DependencyInjection.ServiceLookup.ServiceProviderEngineScope serviceProviderEngineScope)	Unknown
 	Microsoft.Extensions.DependencyInjection.dll!Microsoft.Extensions.DependencyInjection.ServiceLookup.ServiceProviderEngineScope.GetService(System.Type serviceType)	Unknown
 	Microsoft.Extensions.DependencyInjection.Abstractions.dll!Microsoft.Extensions.DependencyInjection.ServiceProviderServiceExtensions.GetService<Microsoft.Extensions.Options.IOptions<Microsoft.ApplicationInsights.Extensibility.TelemetryConfiguration>>(System.IServiceProvider provider)	Unknown
 	Microsoft.ApplicationInsights.AspNetCore.dll!Microsoft.Extensions.DependencyInjection.ApplicationInsightsExtensions.AddApplicationInsightsTelemetry.AnonymousMethod__13_2(System.IServiceProvider provider)	Unknown
 	Microsoft.Extensions.DependencyInjection.dll!Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitFactory(Microsoft.Extensions.DependencyInjection.ServiceLookup.FactoryCallSite factoryCallSite, Microsoft.Extensions.DependencyInjection.ServiceLookup.ServiceProviderEngineScope scope)	Unknown
 	Microsoft.Extensions.DependencyInjection.dll!Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteVisitor<Microsoft.Extensions.DependencyInjection.ServiceLookup.ServiceProviderEngineScope, object>.VisitCallSite(Microsoft.Extensions.DependencyInjection.ServiceLookup.IServiceCallSite callSite, Microsoft.Extensions.DependencyInjection.ServiceLookup.ServiceProviderEngineScope argument)	Unknown
 	Microsoft.Extensions.DependencyInjection.dll!Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitScoped(Microsoft.Extensions.DependencyInjection.ServiceLookup.ScopedCallSite scopedCallSite, Microsoft.Extensions.DependencyInjection.ServiceLookup.ServiceProviderEngineScope scope)	Unknown
 	Microsoft.Extensions.DependencyInjection.dll!Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitSingleton(Microsoft.Extensions.DependencyInjection.ServiceLookup.SingletonCallSite singletonCallSite, Microsoft.Extensions.DependencyInjection.ServiceLookup.ServiceProviderEngineScope scope)	Unknown
 	Microsoft.Extensions.DependencyInjection.dll!Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteVisitor<Microsoft.Extensions.DependencyInjection.ServiceLookup.ServiceProviderEngineScope, object>.VisitCallSite(Microsoft.Extensions.DependencyInjection.ServiceLookup.IServiceCallSite callSite, Microsoft.Extensions.DependencyInjection.ServiceLookup.ServiceProviderEngineScope argument)	Unknown
 	Microsoft.Extensions.DependencyInjection.dll!Microsoft.Extensions.DependencyInjection.ServiceLookup.DynamicServiceProviderEngine.RealizeService.AnonymousMethod__0(Microsoft.Extensions.DependencyInjection.ServiceLookup.ServiceProviderEngineScope scope)	Unknown
 	Microsoft.Extensions.DependencyInjection.dll!Microsoft.Extensions.DependencyInjection.ServiceLookup.ServiceProviderEngine.GetService(System.Type serviceType, Microsoft.Extensions.DependencyInjection.ServiceLookup.ServiceProviderEngineScope serviceProviderEngineScope)	Unknown
 	Microsoft.Extensions.DependencyInjection.dll!Microsoft.Extensions.DependencyInjection.ServiceLookup.ServiceProviderEngine.GetService(System.Type serviceType)	Unknown
 	Microsoft.Extensions.DependencyInjection.dll!Microsoft.Extensions.DependencyInjection.ServiceProvider.GetService(System.Type serviceType)	Unknown
 	Microsoft.Extensions.DependencyInjection.Abstractions.dll!Microsoft.Extensions.DependencyInjection.ServiceProviderServiceExtensions.GetService<Microsoft.ApplicationInsights.Extensibility.TelemetryConfiguration>(System.IServiceProvider provider)	Unknown
 	Microsoft.ApplicationInsights.AspNetCore.dll!Microsoft.ApplicationInsights.AspNetCore.ApplicationInsightsStartupFilter.Configure.AnonymousMethod__0(Microsoft.AspNetCore.Builder.IApplicationBuilder app)	Unknown
 	Microsoft.AspNetCore.Server.IISIntegration.dll!Microsoft.AspNetCore.Server.IISIntegration.IISSetupFilter.Configure.AnonymousMethod__0(Microsoft.AspNetCore.Builder.IApplicationBuilder app)	Unknown
 	Microsoft.AspNetCore.dll!Microsoft.AspNetCore.HostFilteringStartupFilter.Configure.AnonymousMethod__0(Microsoft.AspNetCore.Builder.IApplicationBuilder app)	Unknown
 	Microsoft.AspNetCore.Hosting.dll!Microsoft.AspNetCore.Hosting.Internal.AutoRequestServicesStartupFilter.Configure.AnonymousMethod__0(Microsoft.AspNetCore.Builder.IApplicationBuilder builder)	Unknown
 	Microsoft.AspNetCore.Hosting.dll!Microsoft.AspNetCore.Hosting.Internal.WebHost.BuildApplication()	Unknown
 	Microsoft.AspNetCore.Hosting.dll!Microsoft.AspNetCore.Hosting.Internal.WebHost.StartAsync(System.Threading.CancellationToken cancellationToken)	Unknown
 	System.Private.CoreLib.dll!System.Runtime.CompilerServices.AsyncMethodBuilderCore.Start<Microsoft.AspNetCore.Hosting.Internal.WebHost.<StartAsync>d__26>(ref Microsoft.AspNetCore.Hosting.Internal.WebHost.<StartAsync>d__26 stateMachine)	Unknown
 	Microsoft.AspNetCore.Hosting.dll!Microsoft.AspNetCore.Hosting.Internal.WebHost.StartAsync(System.Threading.CancellationToken cancellationToken)	Unknown
 	Microsoft.AspNetCore.Hosting.dll!Microsoft.AspNetCore.Hosting.WebHostExtensions.RunAsync(Microsoft.AspNetCore.Hosting.IWebHost host, System.Threading.CancellationToken token, string shutdownMessage)	Unknown
 	System.Private.CoreLib.dll!System.Runtime.CompilerServices.AsyncMethodBuilderCore.Start<Microsoft.AspNetCore.Hosting.WebHostExtensions.<RunAsync>d__5>(ref Microsoft.AspNetCore.Hosting.WebHostExtensions.<RunAsync>d__5 stateMachine)	Unknown
 	Microsoft.AspNetCore.Hosting.dll!Microsoft.AspNetCore.Hosting.WebHostExtensions.RunAsync(Microsoft.AspNetCore.Hosting.IWebHost host, System.Threading.CancellationToken token, string shutdownMessage)	Unknown
 	Microsoft.AspNetCore.Hosting.dll!Microsoft.AspNetCore.Hosting.WebHostExtensions.RunAsync(Microsoft.AspNetCore.Hosting.IWebHost host, System.Threading.CancellationToken token)	Unknown
 	System.Private.CoreLib.dll!System.Runtime.CompilerServices.AsyncMethodBuilderCore.Start<Microsoft.AspNetCore.Hosting.WebHostExtensions.<RunAsync>d__4>(ref Microsoft.AspNetCore.Hosting.WebHostExtensions.<RunAsync>d__4 stateMachine)	Unknown
 	Microsoft.AspNetCore.Hosting.dll!Microsoft.AspNetCore.Hosting.WebHostExtensions.RunAsync(Microsoft.AspNetCore.Hosting.IWebHost host, System.Threading.CancellationToken token)	Unknown
 	Microsoft.AspNetCore.Hosting.dll!Microsoft.AspNetCore.Hosting.WebHostExtensions.Run(Microsoft.AspNetCore.Hosting.IWebHost host)	Unknown
 	SignService.dll!SignService.Program.Main(string[] args) Line 13	C#
 	[Native to Managed Transition]	
 	hostpolicy.dll!00007ffec842fe50()	Unknown
 	hostpolicy.dll!00007ffec843064d()	Unknown
 	hostfxr.dll!00007ffef47adea8()	Unknown
 	hostfxr.dll!00007ffef47b6fed()	Unknown
 	hostfxr.dll!00007ffef47b54b2()	Unknown
 	hostfxr.dll!00007ffef47b4426()	Unknown
 	hostfxr.dll!00007ffef47aeae5()	Unknown
 	dotnet.exe!00007ff7e6fa9d75()	Unknown
 	dotnet.exe!00007ff7e6faf839()	Unknown
 	kernel32.dll!00007fff32ad7e94()	Unknown
 	ntdll.dll!00007fff33c5a251()	Unknown
```

**Comment by onovotny** [comment]
Bump?

**Comment by sywhang** [comment]
Not sure why this got closed... Leaving the issue open for backporting the fix to 2.2

**Comment by sywhang** [comment]
Fix got merged to both 2.2 / master.
